tests/test_utils/alloc.py

```diff
@@ -0,0 +1,88 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gc
+import pytest
+import tracemalloc
+
+
+class MallocInstant:
+    """
+    This class wraps allocation tracing in a way that pretends to be a
+    clock, that is, one can ask for "now" snapshots, and then
+    difference them to work out the "elapsed" allocations (that is,
+    the total size difference) between them.
+
+    The __sub__ (-) operator counts both allocations (positive) and
+    frees (negative) between the two "MallocInstant"s, and so gives
+    the delta in the size of the allocations, not the total size of
+    new allocations.
+    """
+
+    def __init__(self, snapshot: tracemalloc.Snapshot):
+        self._snapshot = snapshot
+
+    def __sub__(self, other: "MallocInstant") -> int:
+        diff = self._snapshot.compare_to(other._snapshot, "lineno")
+        return sum(elem.size_diff for elem in diff)
+
+
+def snapshot(gc_generation=0) -> MallocInstant:
+    """
+    Take a snapshot of the current "malloc instant", similar to an allocation version of time.perf_counter() (etc.).
+    """
+    if gc_generation is not None:
+        gc.collect(gc_generation)
+    return MallocInstant(tracemalloc.take_snapshot())
+
+
+@pytest.fixture
+def allocation_benchmark(request, benchmark):
+    # make sure the user specified the "snapshot" timer
+    marker = request.node.get_closest_marker("benchmark")
+    options = dict(marker.kwargs) if marker else {}
+    correct_timer = "timer" in options and options["timer"] is snapshot
+    if not correct_timer:
+        raise ValueError(
+            "allocation_benchmark fixture can only be used in functions with @pytest.mark.benchmark(..., timer=%s.%s, ...)"
+            % (__name__, snapshot.__name__)
+        )
+
+    benchmark.extra_info["allocation_benchmark"] = True
+
+    def run_it(f):
+        result = benchmark.pedantic(f, iterations=1, rounds=20, warmup_rounds=3)
+        if result is None:
+            raise ValueError(
+                "benchmark function returned None: allocation benchmarking is only reliable if the object(s) of interest is created inside and returned from the function being benchmarked"
+            )
+
+    # Running without a GC for an memory benchmark? This ensures that all objects created in the
+    # benchmarked function get placed into the young generation (0), and so cleaning those up to
+    # leave only the long-lived objects from each function execution is just a `gc.collect(0)` which
+    # is very fast (much faster than `gc.collect()`). For measuring peak memory use this gives the
+    # worst-case peak: the case when no GC runs at all during the benchmarked-function and so only
+    # obvious deallocations occur.
```

Comment:
> You are disabling the GC to run the tests, is this the same as "running without a GC"?

Yes, I was taking "disabled GC" = "no GC". I'll rephrase because I guess it's a bit sloppy.

> Are you saying if you perform a gc.collect(0) we could measure the worst case peak?

Ah, no, that's meant to be read as:

> Disabling GC is beneficial for measuring the peak memory use too, if that was implemented (e.g. via `tracemalloc.get_traced_memory`), as it gives us the worst-case peak: the case when no GC runs at all during the benchmarked-function and so only obvious deallocations occur.

Does that make more sense? I'll rephrase (e.g. copy that in), or even remove, if we're not going to do anything with the peak memory use.

Reply:
Why do you mention gc.collect(0) s faster when GC is disabled, but then don't use this?  The second paragraph is great 👍

Reply:
> Why do you mention gc.collect(0) s faster when GC is disabled, but then don't use this?

Ah, the key here is: disabling GC allows us to use `gc.collect(0)`.

Collecting just the young generation with `gc.collect(0)` is always faster than collecting all generations with `gc.collect()`, whether or not GC is stopped from running automatically (which is what `gc.disable()` is doing). By doing the latter, GC never runs without us asking for it, and so objects never move from the young generation (gen 0) where they are first allocated to gen 1 (or from gen 1 to gen 2).

In addition, this code does use `gc.collect(0)` (in `snapshot`).

Does that clarify the motivation better? I'll think how I can rephrase the comment.

Reply:
I've rewritten the comment. Maybe that's clearer?

Reply:
Yes, the description is super clear now, but still in the final line you are using `gc.collect()` not `gc.collect(0)` so this comment is informative about gc in Python but doesn't seem to match what you actually do, or am I missing something?

Reply:
Ah, no, that is just cleaning up everything after the benchmark has completely run; this comment ties into `snapshot`'s `gc.collect` call.